term/src/cache.rs

```diff
@@ -0,0 +1,217 @@
+            RawValue(iri, None) => Iri::new_unchecked(iri).into(),
+        }
+    }
+
+    /// Number of cached data.
+    pub fn len(&self) -> usize {
+        self.0.len()
+    }
+
+    /// Get a reference to cached `TermData`. If it was not cached already a
+    /// new instance is cached.
+    pub fn get_term_data<TD>(&mut self, txt: TD) -> CD
+    where
+        TD: TermData + IntoData<CD>,
+    {
+        if let Some(term_data) = self.0.get(txt.as_ref()) {
+            term_data
+        } else {
+            let term_data = txt.into_data();
+            self.0.insert(Clone::clone(&term_data));
+            term_data
+        }
+    }
+
+    /// Removes unused elements from the cache.
+    pub fn shrink_to_fit(&mut self) {
+        self.0.shrink_to_fit()
+    }
+
+    /// Build a new IRI.
+    pub fn iri<TD>(&mut self, iri: TD) -> Result<Term<CD>>
+    where
+        TD: TermData + IntoData<CD>,
+    {
+        Term::new_iri(self.get_term_data(iri))
+    }
+
+    /// Build a new IRI from namespace and suffix.
+    pub fn iri_suffixed<TD1, TD2>(&mut self, ns: TD1, suffix: TD2) -> Result<Term<CD>>
+    where
+        TD1: TermData + IntoData<CD>,
+        TD2: TermData + IntoData<CD>,
+    {
+        Term::new_iri_suffixed(self.get_term_data(ns), self.get_term_data(suffix))
+    }
+
+    /// Build a new blank node with the given ID.
+    pub fn bnode<TD>(&mut self, id: TD) -> Result<Term<CD>>
+    where
+        TD: TermData + IntoData<CD>,
+    {
+        Term::new_bnode(self.get_term_data(id))
+    }
+
+    /// Build a new, typed literal.
+    pub fn literal_dt<TD, T>(&mut self, txt: TD, dt: T) -> Result<Term<CD>>
+    where
+        TD: TermData,
+        T: TTerm,
+        CD: for<'a> FromData<&'a str> + FromData<TD> + for<'a> From<&'a str>,
+    {
+        if TermKind::Iri == dt.kind() {
+            Term::new_literal_dt(self.get_term_data(txt), self.iri_from_raw(dt.value_raw()))
+        } else {
+            Err(TermError::NotAnIri(dt.value().to()))
+        }
+    }
+
+    /// Build a new, language-tagged literal.
+    pub fn literal_lang<TD1, TD2>(&mut self, txt: TD1, lang: TD2) -> Result<Term<CD>>
+    where
+        TD1: TermData + IntoData<CD>,
+        TD2: TermData + IntoData<CD>,
+    {
+        Term::new_literal_lang(self.get_term_data(txt), self.get_term_data(lang))
+    }
+
+    /// Build a new variable with the given ID.
+    pub fn variable<TD>(&mut self, id: TD) -> Result<Term<CD>>
+    where
+        TD: TermData + IntoData<CD>,
+    {
+        Term::new_variable(self.get_term_data(id))
+    }
+
+    /// Clone any `Term`.
+    pub fn clone_term<TD: TermData>(&mut self, other: &Term<TD>) -> Term<CD>
+    where
+        CD: for<'a> FromData<&'a str>,
+    {
+        other.clone_map(|td| self.get_term_data(td))
+    }
+
+    /// Clone anything that implements the `TTerm` trait.
+    pub fn clone_dyn_term(&mut self, other: &dyn TTerm) -> Term<CD>
```

Comment:
couldn't this be polymorphic instead?
```suggestion
    pub fn clone_tterm<T: TTerm>(&mut self, other: &T) -> Term<CD>
```